Add cli-ssh-oauth-device-grant post (#865)
oktadeveloper/okta-blog: _source/_posts/2021-09-20-cli-ssh-oauth-device-grant.md

---
layout: blog_post
title: "Command Line Login with the OAuth Device Grant"
author: [huan-liu,  lars-johansen]
by: internal-contributor
communities: [mobile]
description: "Tutorial: Learn how to log into input limited devices like TVs and terminals using the OAuth 2.0 Device Grant!"
tags: [oauth, oauth2, security, mobile, oauth2-device-flow, device-flow, ssh, c]
tweets:
- "Here's how to do command line login for input-constrained devices, using the OAuth 2.0 Device Authorization Grant. Stream on!"
- "Command line sign-in to input-constrained devices—check it out and tell us what you think!"
#image: blog/cli-ssh-oauth-device-grant/social.jpg
type: conversion
---

Even if you've never heard of the OAuth 2.0 Device Authorization Grant, formerly known as the Device Flow, you've probably seen it in action on a smart TV or other streaming device. The Device Authorization Grant is commonly used on devices with limited keyboard input ability. The process allows you to log in to an account more easily via a separate browser on your laptop, mobile phone, or other companion rich-input device.

> For instance, if AtkoTV is launching a streaming service; its streaming Apple TV app might present you with something like the sign-in screen below. Instead of struggling to input anything on your TV, you'd log in on your laptop or mobile phone to complete the sign-in process.

Reviewer: ```suggestion
For instance, if AtkoTV is launching a streaming service, its streaming Apple TV app might present you with something like the sign-in screen below. Instead of struggling to input anything on your TV, you'd log in on your laptop or mobile phone to complete the sign-in process.
```

  --data 'client_id=0oa15wulqt5yqD9FP5d7&scope=openid profile offline_access'
```

We will get back a response:

```json
{
  "device_code": "df1b060b-4e36-4bbe-98aa-5dcb11909f5f",
  "user_code": "DRTDNTJC",
  "verification_uri": "https://dev-133337.okta.com/activate",
  "expires_in": 600,
  "interval": 5
}
```

In the second step, to poll for the access token, we issue the following command:

```sh
curl -X POST \
  --url https://dev-133337.com/oauth2/v1/token \
  --data 'device_code=df1b060b-4e36-4bbe-98aa-5dcb11909f5f&grant_type=urn:ietf:params:oauth:grant-type:device_code&client_id=0oa15wulqt5yqD9FP5d7'
```

If the user has not logged in, the response looks like this:

```json
{
  "error": "authorization_pending",
  "error_description": "The device authorization is pending. Please try again later."
}
```

Otherwise, we'll get a response back with the `access_token`, `refresh_token` and `id_token`.

```json
{
  "token_type": "Bearer",
  "expires_in": 3600,
  "access_token": "....",
  "scope": "openid profile offline_access",
  "refresh_token": "....",
  "id_token": "...."
}
```

## How to Enable the Device Grant in Okta

> <!-- Remove following section after features is GA →

Reviewer: ```suggestion
<!-- Remove following section after features is GA -->
```

```
# @include common-auth             # Disable password authentication
auth    required    deviceflow.so  # register custom authentication module
```

Next we'll edit the `/etc/ssh/sshd_config` file to make sure that challenge-response authentication is enabled, using PAM.  You'll see a line starting with `ChallengeResponseAuthentication`. The default is `no`. Change this to `yes`:

```
ChallengeResponseAuthentication yes
```

Later on in the same file, make sure that "UsePAM" is set to yes (it is the default) as follows:

```
UsePAM yes
```

Next we'll develop the authentication module. To make sure the necessary libraries are installed, we run the following commands:

```sh
apt-get update
apt-get install build-essential libpam0g-dev libcurl4-openssl-dev libqrencode-dev libssl-dev
```

The full source code for the PAM module is available on [GitHub](https://github.com/oktadev/okta-ssh-oauth-example). For a PAM module, we need to define two interfaces:

```c
PAM_EXTERN int pam_sm_setcred( pam_handle_t *pamh, int flags, int argc, const char **argv)
PAM_EXTERN int pam_sm_authenticate( pam_handle_t *pamh, int flags,int argc, const char **argv )
```

> The second interface handles the authentication flow, which holds all our authentication logic. The code is self explanatory, so we'll only cover the high-level steps:

Reviewer 2: ```suggestion
The second interface handles the authentication flow, which holds all our authentication logic. The code is self-explanatory, so we'll only cover the high-level steps:
```
needs hypen

> 1. First, plug in your org's urls in `deviceflow.c`. Change the url based on your org name and whether you are using the org Authentication Server or a custom Authentication Server. Also, define the client ID for the Native App where you've enabled the Device Authorization grant type.

Reviewer 2: ```suggestion
1. First, plug in your org's URLs in `deviceflow.c`. Change the URL based on your org name and whether you are using the org Authentication Server or a custom Authentication Server. Also, define the client ID for the Native App where you've enabled the Device Authorization grant type.
```
url-> URL per dev docs style guide

   ```c
   #define DEVICE_AUTHORIZE_URL  "https://dev-133337.okta.com/oauth2/v1/device/authorize"
   #define TOKEN_URL "https://dev-133337.okta.com/oauth2/v1/token"
   #define CLIENT_ID "0oa15wulqt5yqD9FP5d7"
   ```

> 2. We first issue a POST call to our `authorize` endpoint on the Authorization Server. From the POST result, we can extract the `device_code` and `user_code`. The POST call is abstracted away as a call to a helper function `issuePost` which takes the URL as the first parameter and the POST data as the second parameter:

Reviewer 2: ```suggestion
2. We first issue a POST call to our `authorize` endpoint on the Authorization Server. From the POST result, we can extract the `device_code` and `user_code`. The POST call is abstracted away as a call to a helper function `issuePost`, which takes the URL as the first parameter and the POST data as the second parameter:
```
comma precedes "which"

   ```c
   sprintf(postData, "client_id=%s&scope=openid profile offline_access", CLIENT_ID);
   issuePost(DEVICE_AUTHORIZE_URL, postData);
   ```

    The helper function uses libcurl to send the POST request:

   ```c
   void issuePost(char * url, char * data) {
      /* send all data to this callback function  */
      curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
      /* we pass our 'chunk' struct to the callback function */
      curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&chunk);
      
      curl_easy_setopt(curl, CURLOPT_URL, url ) ;
      curl_easy_setopt(curl, CURLOPT_POST, 1);  /* this is a POST */
      curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
      int res = curl_easy_perform( curl ) ;
   }
   ```

    The `WriteMemoryCallback` function saves the callback results in a `chunk` buffer. Then, we use a helper function `getValueForKey` to extract the value fields for keys `user_code`, `device_code`, and `verification_uri`. 

> 

Reviewer 2: ```suggestion

```
will this add a linebreak? it looks really tight btwn steps 2 & 3.

Author: Nope, I just tested it locally, I think it's a CSS issue, we _could_ probably add a bit of padding to the whole site between lists. But let's deal with that outside of this PR

> 3. We then prompt the user to login to the activate endpoint. For a more pleasant user experience, we incorporate the URL inside a QRCode. We leveraged the code from Y2Z/qr:  Text-based QR Code® generator to generate ASCII output so that it will output properly in the command shell. 

Reviewer: ```suggestion
3. We then prompt the user to login to the `activate` endpoint. For a more pleasant user experience, we incorporate the URL inside a QR code. We leveraged the code from [Y2Z/qr](https://github.com/Y2Z/qr) to generate ASCII output so that it will output properly in the command shell. 
```
add code formatting to activate endpoint for consistency with earlier authorize endpoint, add space & lowercase in "QR code"

    ```c
    char * qrc = getQR(activateUrl);
    sprintf(prompt_message, 
      "\n\nPlease login at %s or scan the QRCode below:\nThen input code %s\n\n%s", 
      activateUrl, usercode, qrc);
    free(qrc);
    sendPAMMessage(pamh, prompt_message);
    ```

    Since this is a PAM module, we cannot just print to screen directly. The `sendPAMMessage` helper function uses the PAM module's interface function to send messages to the user:
    
   ```c
   void
   sendPAMMessage(pam_handle_t *pamh, char * prompt_message) {
      int retval;
      
      struct pam_message msg[1],*pmsg[1];
      struct pam_response *resp;
      struct pam_conv *conv ;
      
      pmsg[0] = &msg[0] ;
      msg[0].msg_style = PAM_TEXT_INFO ; 
      msg[0].msg = prompt_message;
      
      resp = NULL ;
      
      retval = pam_get_item( pamh, PAM_CONV, (const void **) &conv ) ;  
      if( retval==PAM_SUCCESS ) {
         retval = conv->conv( 1, (const struct pam_message **) pmsg, &resp,  conv->appdata_ptr ) ; 
      }
      if( resp ) {
         free( resp );
      }
   }
   ```

>    In above, we set the communication type to `PAM_TEXT_INFO` to indicate that we want to send a message to the user. In the `pam_get_item` call, we retrieve the PAM conversation module `conv`, and in the following line, we just invoke the `conv` module, which writes the message to the user console.

Reviewer 2: ```suggestion
   In the above, we set the communication type to `PAM_TEXT_INFO` to indicate that we want to send a message to the user. In the `pam_get_item` call, we retrieve the PAM conversation module `conv`, and in the following line, we just invoke the `conv` module, which writes the message to the user console.
```
added "the" to correct the expression

4. We periodically issue a POST call to the token endpoint. If the POST result gives a token back, we return `PAM_SUCCESS` to allow the user to log in, or eventually, we return `PAM_AUTH_ERR` to signal authentication failure and deny user access.

   ```c
   sprintf(postData, "device_code=%s&grant_type=urn:ietf:params:oauth:grant-type:device_code&client_id=%s", devicecode, CLIENT_ID);

   while (waitingForActivate) {

>       issuePost(tokenUrl, postData);

Reviewer: ```suggestion
      issuePost(TOKEN_URL, postData);
```

      // ...
   }
   ```

5. The POST return includes various JWT tokens. From the `id_token`, we know exactly who logged in, and we optionally print a customized welcome message.

To compile all our code (`deviceflow.c` and `qr.c`), we run:

```sh
gcc -fPIC -c deviceflow.c qr.c
sudo ld -x --shared -o /lib/security/deviceflow.so deviceflow.o qr.o \
        -lm -lqrencode -lcurl -lssl -lcrypto
```

> The second command installs the compiled PAM module under /lib/security, which is where PAM looks for additional plugin modules.

Reviewer 2: ```suggestion
The second command installs the compiled PAM module under `/lib/security`, which is where PAM looks for additional plugin modules.
```
use code formatting
